src/VisualStudio/Core/Def/Implementation/ProjectSystem/RunningDocumentTableEventTracker.cs

```diff
@@ -0,0 +1,192 @@
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    /// <summary>
+    /// Class to register with the RDT and forward RDT events.
+    /// Handles common conditions and delegates implementation to the <see cref="IRunningDocumentTableEventListener"/>
+    /// </summary>
+    internal sealed class RunningDocumentTableEventTracker : IVsRunningDocTableEvents3, IDisposable
+    {
+        private bool _isDisposed = false; // To detect redundant calls
+
+        private readonly ForegroundThreadAffinitizedObject _foregroundAffinitization;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly IVsRunningDocumentTable4 _runningDocumentTable;
+        private readonly IRunningDocumentTableEventListener _listener;
+        private uint _runningDocumentTableEventsCookie;
+
+        public RunningDocumentTableEventTracker(IThreadingContext threadingContext, IVsEditorAdaptersFactoryService editorAdaptersFactoryService, IVsRunningDocumentTable4 runningDocumentTable,
+            IRunningDocumentTableEventListener listener)
+        {
+            Contract.ThrowIfNull(threadingContext);
+            Contract.ThrowIfNull(editorAdaptersFactoryService);
+            Contract.ThrowIfNull(runningDocumentTable);
+            Contract.ThrowIfNull(listener);
+
+            _foregroundAffinitization = new ForegroundThreadAffinitizedObject(threadingContext, assertIsForeground: true);
+            _runningDocumentTable = runningDocumentTable;
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            _listener = listener;
+
+            ((IVsRunningDocumentTable)_runningDocumentTable).AdviseRunningDocTableEvents(this, out _runningDocumentTableEventsCookie);
+        }
+
+        public int OnAfterFirstDocumentLock(uint docCookie, uint dwRDTLockType, uint dwReadLocksRemaining, uint dwEditLocksRemaining)
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public int OnBeforeLastDocumentUnlock(uint docCookie, uint dwRDTLockType, uint dwReadLocksRemaining, uint dwEditLocksRemaining)
+        {
+            if (dwReadLocksRemaining + dwEditLocksRemaining == 0)
+            {
+                if (CheckPreconditions(docCookie))
+                {
+                    _listener.OnCloseDocument(docCookie, _runningDocumentTable.GetDocumentMoniker(docCookie));
+                }
+            }
+
+            return VSConstants.S_OK;
+        }
+
+        public int OnAfterSave(uint docCookie)
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public int OnAfterAttributeChange(uint docCookie, uint grfAttribs)
+        {
+            return VSConstants.S_OK;
```

Comment:
This should return E_NOTIMPL. This is a general contract (which isn't always followed by the VS shell...) where if you return E_NOTIMPL for events like this it gets the hint that we don't care and won't call us again.

Reply:
Good to know - Misc workspace returns OK, VS workspace returns E_NOTIMPL, so I picked OK.  Will change.

https://github.com/dotnet/roslyn/blob/master/src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs#L120